# Imports

In [58]:
import pandas as pd
pd.options.mode.chained_assignment = None

import spacy
from spacy.lang.en.stop_words import STOP_WORDS
import en_core_web_sm

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import accuracy_score
from sklearn.base import TransformerMixin
from sklearn.svm import LinearSVC
from sklearn.pipeline import Pipeline

### Setup Spacy NLP

In [59]:
nlp = en_core_web_sm.load()
spacy_tokenizer = nlp.tokenizer

### Load Dataset(s)

In [60]:
df = pd.read_csv('data/train.csv')

### Explore & Clean Data

In [61]:
df.head(25)

2401  Borderlands  1 im getting on borderlands and i will murder you all ,
0   2401  Borderlands  1  I am coming to the borders and I will kill you...   
1   2401  Borderlands  1  im getting on borderlands and i will kill you ...   
2   2401  Borderlands  1  im coming on borderlands and i will murder you...   
3   2401  Borderlands  1  im getting on borderlands 2 and i will murder ...   
4   2401  Borderlands  1  im getting into borderlands and i can murder y...   
5   2402  Borderlands  1  So I spent a few hours making something for fu...   
6   2402  Borderlands  1  So I spent a couple of hours doing something f...   
7   2402  Borderlands  1  So I spent a few hours doing something for fun...   
8   2402  Borderlands  1  So I spent a few hours making something for fu...   
9   2402  Borderlands  1  2010 So I spent a few hours making something f...   
10  2402  Borderlands  1                                                was   
11  2403  Borderlands  0  Rock-Hard La Varlope, RARE & POWERFUL, HANDSOM...   
12  2403  Borderlands  0  Rock-Hard La Varlope, RARE & POWERFUL, HANDSOM...   
13  2403  Borderlands  0  Rock-Hard La Varlope, RARE & POWERFUL, HANDSOM...   
14  2403  Borderlands  0  Rock-Hard La Vita, RARE BUT POWERFUL, HANDSOME...   
15  2403  Borderlands  0  Live Rock - Hard music La la Varlope, RARE & t...   
16  2403  Borderlands  0  I-Hard like me, RARE LONDON DE, HANDSOME 2011,...   
17  2404  Borderlands  1  that was the first borderlands session in a lo...   
18  2404  Borderlands  1  this was the first Borderlands session in a lo...   
19  2404  Borderlands  1  that was the first borderlands session in a lo...   
20  2404  Borderlands  1  that was the first borderlands session in a lo...   
21  2404  Borderlands  1  that I was the first real borderlands session ...   
22  2404  Borderlands  1  that was the first borderlands session in a ho...   
23  2405  Borderlands -1  the biggest dissappoinment in my life came out...   
24  2405  Borderlands -1  The biggest disappointment of my life came a y...

In [62]:
df.columns = ['id', 'entity', 'sentiment', 'review']

In [63]:
train = df.drop(columns=['id', 'entity'])

In [64]:
train.head(25)

sentiment                                             review
0           1  I am coming to the borders and I will kill you...
1           1  im getting on borderlands and i will kill you ...
2           1  im coming on borderlands and i will murder you...
3           1  im getting on borderlands 2 and i will murder ...
4           1  im getting into borderlands and i can murder y...
5           1  So I spent a few hours making something for fu...
6           1  So I spent a couple of hours doing something f...
7           1  So I spent a few hours doing something for fun...
8           1  So I spent a few hours making something for fu...
9           1  2010 So I spent a few hours making something f...
10          1                                                was
11          0  Rock-Hard La Varlope, RARE & POWERFUL, HANDSOM...
12          0  Rock-Hard La Varlope, RARE & POWERFUL, HANDSOM...
13          0  Rock-Hard La Varlope, RARE & POWERFUL, HANDSOM...
14          0  Rock-Hard La Vita, RARE BUT POWERFUL, HANDSOME...
15          0  Live Rock - Hard music La la Varlope, RARE & t...
16          0  I-Hard like me, RARE LONDON DE, HANDSOME 2011,...
17          1  that was the first borderlands session in a lo...
18          1  this was the first Borderlands session in a lo...
19          1  that was the first borderlands session in a lo...
20          1  that was the first borderlands session in a lo...
21          1  that I was the first real borderlands session ...
22          1  that was the first borderlands session in a ho...
23         -1  the biggest dissappoinment in my life came out...
24         -1  The biggest disappointment of my life came a y...

In [65]:
train.value_counts('sentiment')

sentiment
 0    31308
-1    22542
 1    20831
dtype: int64

In [66]:
train = train[train['sentiment'] != 0]

In [67]:
train.value_counts('sentiment')

sentiment
-1    22542
 1    20831
dtype: int64

In [68]:
train['sentiment'].mask(train['sentiment'] == -1, 0, inplace=True)

In [69]:
train.value_counts('sentiment')

sentiment
0    22542
1    20831
dtype: int64

In [70]:
train.isna().sum()

sentiment      0
review       361
dtype: int64

In [71]:
train.dropna(axis=0, inplace=True)

In [72]:
train.isna().sum()

sentiment    0
review       0
dtype: int64

In [73]:
stopwords = list(STOP_WORDS)
stopwords

['because',
 'out',
 'her',
 "'s",
 'does',
 'even',
 'will',
 'been',
 'any',
 'name',
 'same',
 "'d",
 'n’t',
 'by',
 'seeming',
 'unless',
 'them',
 'nowhere',
 'when',
 'towards',
 'was',
 'to',
 "'re",
 'your',
 'former',
 'who',
 'between',
 'hereby',
 'seem',
 'those',
 'part',
 'someone',
 'myself',
 'however',
 'ten',
 'have',
 'made',
 'other',
 'my',
 'across',
 'amongst',
 'empty',
 'yours',
 'after',
 '‘s',
 'besides',
 'much',
 'one',
 'for',
 'further',
 'top',
 'it',
 '‘d',
 '’m',
 'again',
 'last',
 'few',
 'anyone',
 'before',
 'thru',
 'many',
 'front',
 'others',
 'regarding',
 'during',
 '‘ve',
 'there',
 'hereupon',
 'two',
 'nevertheless',
 'whole',
 'sometimes',
 'had',
 'elsewhere',
 'should',
 'we',
 'us',
 'into',
 'least',
 'if',
 'none',
 'might',
 'hence',
 'though',
 'a',
 'above',
 'still',
 'each',
 'thereafter',
 'you',
 'formerly',
 'say',
 'thence',
 'nor',
 'perhaps',
 'else',
 'whoever',
 'only',
 'without',
 'whither',
 'do',
 'often',
 'whereupon

In [74]:
def clean_text(s):
    split_text = s.split()
    cleaned_text = [word for word in split_text if word not in stopwords]
    final_text = ' '.join(cleaned_text)
    return final_text

In [75]:
text = "So I spent a couple of hours doing something for fun... If you don't know that I'm a huge @ Borderlands fan and Maya is one of my favorite characters, I decided to make a wallpaper for my PC.. Her..."

clean_text(text)

"So I spent couple hours fun... If don't know I'm huge @ Borderlands fan Maya favorite characters, I decided wallpaper PC.. Her..."

In [76]:
train['review'] = train['review'].apply(clean_text)

In [77]:
train.head(25)

sentiment                                             review
0           1                       I coming borders I kill all,
1           1                   im getting borderlands kill all,
2           1                  im coming borderlands murder all,
3           1               im getting borderlands 2 murder all,
4           1                 im getting borderlands murder all,
5           1  So I spent hours making fun. . . If don't know...
6           1  So I spent couple hours fun... If don't know I...
7           1  So I spent hours fun... If don't know I'm HUGE...
8           1  So I spent hours making fun. . . If don't know...
9           1  2010 So I spent hours making fun. . . If don't...
10          1                                                   
17          1  borderlands session long time actually satisfy...
18          1  Borderlands session long time actually satisfy...
19          1  borderlands session long time actually satisfy...
20          1  borderlands session long time actually enjoyed...
21          1  I real borderlands session nice long wait time...
22          1  borderlands session hot row actually bad comba...
23          0  biggest dissappoinment life came year ago fuck...
24          0     The biggest disappointment life came year ago.
25          0     The biggest disappointment life came year ago.
26          0  biggest dissappoinment life coming year ago fu...
27          0  For biggest male dissappoinment life came hang...
28          0  biggest dissappoinment life came year ago fuck...
29          1  WE FINISHED BORDERLANDS 3 FINALLY YAS! Thank h...
30          1  WE FINALLY FINALLY FIND BORDERLANDS 3 YES! Tha...

### Create Data Processing Pipeline

In [78]:
def cleaner(df_path):
    initial_df = pd.read_csv(df_path)
    initial_df.columns = ['id', 'entity', 'sentiment', 'review']
    initial_df.drop(columns=['id', 'entity'])
    prepped_df = initial_df[initial_df['sentiment'] != 0]
    prepped_df['sentiment'].mask(prepped_df['sentiment'] == -1, 0, inplace=True)
    prepped_df['review'] = prepped_df['review'].apply(clean_text)
    return prepped_df

#### Utilize data processing to prep test data

In [79]:
test = cleaner('data/test.csv')

In [80]:
test.head(25)

id                  entity  sentiment  \
1    8312               Microsoft          0   
2    4371                   CS-GO          0   
4    6273                    FIFA          0   
5    7925               MaddenNFL          1   
6   11332    TomClancysRainbowSix          1   
7    1107          AssassinsCreed          1   
8    2069              CallOfDuty          0   
9    3185                   Dota2          1   
10   1172          AssassinsCreed          1   
11  11783                 Verizon          0   
13   8431                   NBA2K          0   
14   9135                  Nvidia          1   
15   4822     GrandTheftAuto(GTA)          1   
16   3068                   Dota2          0   
17  10537  RedDeadRedemption(RDR)          1   
18   8056               Microsoft          0   
19   2131              CallOfDuty          0   
22   4038                   CS-GO          0   
25  11995                 Verizon          0   
28   2419             Borderlands          0   
29   8857                  Nvidia          1   
30   9704       PlayStation5(PS5)          1   
31   7024         johnson&johnson          0   
32   8842                  Nvidia          1   
33   6650                Fortnite          0   

                                               review  
1   @Microsoft Why I pay WORD functions poorly @Sa...  
2   CSGO matchmaking closet hacking, it's truly aw...  
4   Hi @EAHelp I’ve Madeleine McCann cellar past 1...  
5   Thank @EAMaddenNFL!! New TE Austin Hooper ORAN...  
6   Rocket League, Sea Thieves Rainbow Six: Siege🤔...  
7   ass knee-deep Assassins Creed Odyssey way anyt...  
8   FIX IT JESUS ! Please FIX IT ! What In world g...  
9   The professional dota 2 scene fucking explodin...  
10  Itching assassinate #TCCGif #AssassinsCreedBla...  
11  @FredTJoseph hey fred, Comcast cut cable Veriz...  
13  @NBA2K game sucks... 2 38 seconds left team in...  
14  Congrats NVIDIA NeMo team 1.0.0 release candid...  
15                                      yeah it’s fun  
16                                        fuck life 😆  
17  happy birthday red dead redemption shit change...  
18  What Microsoft hardware & software security - ...  
19             The new @CallofDuty ps5 🔥🔥🔥🔥 Oh God 😭😍  
22  Finally played Rainbow Six Siege time... I adm...  
25  Yo! @Verizon added $120 'fee' account #COVID19...  
28  @Borderlands I submit complaint? Your CEO isn'...  
29  Watching NVIDIA position leading hardware manu...  
30  I don’t looks like Xbox controller y’all anyth...  
31  Johnson & Johnson, knowingly sold baby powder ...  
32        The things I @nvidia 3090... unspeakable! 🧐  
33         Fortnite running like ass.. fps drops wtf?

### Setup sklearn imports

In [92]:
X_train = train['review']
y_train = train['sentiment']

X_test = test['review']
y_test = test['sentiment']

In [82]:
pipe = Pipeline([('vectorizer', TfidfVectorizer()),
                 ('classifier', LinearSVC())])

In [83]:
pipe_model = pipe.fit(X_train, y_train)

In [98]:
predictions = pipe_model.predict(X_test)
accuracy_score(y_test, predictions)

0.9815837937384899

In [99]:
pipe2 = Pipeline([('vectorizer', TfidfVectorizer()),
                 ('classifier', MultinomialNB())])

In [100]:
pipe2_model = pipe2.fit(X_train, y_train)

In [101]:
predictions2 = pipe2_model.predict(X_test)
accuracy_score(y_test, predictions2)

0.9392265193370166

In [88]:
pipe3 = Pipeline([('vectorizer', TfidfVectorizer()),
                 ('classifier', SGDClassifier(loss='hinge', penalty='l2',
                                              alpha=1e-3, random_state=42,
                                              max_iter=5, tol=None))])

In [89]:
pipe3_model = pipe3.fit(X_train, y_train)

In [102]:
predictions3 = pipe3_model.predict(X_test)
accuracy_score(y_test, predictions3)

0.8379373848987108

# Pipe Performance

Pipe #1 Accuracy: 0.9815837937384899
Pipe #2 Accuracy: 0.9392265193370166
Pipe #3 Accuracy: 0.8379373848987108

## Import joblib and save Pipe #1

In [109]:
import joblib

In [112]:
joblib.dump(pipe, 'sk_sentiment.joblib')

['sk_sentiment.joblib']